File: decryptVietnamese.java
public class CheckVietNamese {

  // Sử dụng kiến thức về tiếng việt từ // https://mltav.asn.au/vietnamese/images/documents/Van/wa-van.pdf

  /* 
  Ý tưởng:
   B1: Tạo ra 26 bản decrypt từ ciphertext đầu vào
   B2: Tìm những tổ hợp phụ âm, nguyên âm, ... của tiếng việt trong từng bản decrypt
   B3: Tính điểm (dựa trên độ dài của tổ hợp tìm được) cho từng bản decrypt
   B4: Lấy bản decrypt có số điểm cao nhất.
   */

  static String phuam = "ch,gh,kh,ng,ngh,nh,ph,th,tr,gi,qu";
  static String banNguyenAm = "oa,oe,uy,ue";
  static String nguyenAm = "a,e,i,o,u,y";
  static String vanDonGian =
    "ic,ich,im,in,inh,ip,it,uc,un,ung,ut,um,up,ech,em,en,enh,ep,et,om,on,op,ot,ac,am,an,ang,ap,at,oc,ong,ec,eng,ach,anh";
  static String vanHoaAm = "eo,ao,ai,oi,eu,ia,iu,ui,uu,ua";
  static String vanHopAmLoai1 = "ay,au";
  static String vanHopAmLoai2 =
    "oac,oam,oan,oat,oang,uan,uat,uang,ieu,iec,iem,ien,iep,iet,ieng,yeu,yem,yen,uoi,uoc,uom,,uon,uop,uot,uong";

  public static void main(String[] args) {
    String ciphertext = "Dtys gtpy nly yza wpy nwlddczzx qtwp xl yrfzy, lys nsfa vpe bfl nslj, nzxxpye ctpyr vst yza mlt szln rst nsf eczyr xl yrfzy j efzyr wlx.";
    String ciphertext2 = "Zpuo cplu tbvu sht ihp WOHP JVTTLUA WBISPJ ov alu ahp ihp cpla uhf aybvj kl jhj zpuo cplu rohj ipla zv sbvun zpuo cplu khun sht.";

    System.out.println("1: " + ciphertext);
    System.out.println(CaesarCipher(ciphertext));

    System.out.println("\n\n2: " + ciphertext2);
    System.out.println(CaesarCipher(ciphertext2));
  }

  public static String CaesarCipher(String ciphertext) {
    String result = "";
    int bestScore = 0;

    for (int shift = 0; shift < 26; shift++) {
      String decrypttext = "";

      for (int i = 0; i < ciphertext.length(); i++) {
        char c = ciphertext.charAt(i);
        if (c == ' ' || c == ',' || c == '.') decrypttext += Character.toString(c);
        else if (Character.isUpperCase(c)) {
          char ch = (char)(((int) c + (26 - shift) - 65) % 26 + 65);
          decrypttext += Character.toString(ch);
        } else {
          char ch = (char)(((int) c + (26 - shift) - 97) % 26 + 97);
          decrypttext += Character.toString(ch);
        }
      }

      int score = checkVietNameseNoSpace(decrypttext);
      if (score > bestScore) {
        bestScore = score;
        result = decrypttext;
      }
    }

    return result;
  }

  public static int checkVietNameseNoSpace(String s) {
    int totalScore = 0;
    String str = s.toLowerCase();
    totalScore += GetScore(str, vanHopAmLoai2.split(","));
    totalScore += GetScore(str, vanDonGian.split(","));
    totalScore += GetScore(str, vanHopAmLoai1.split(","));
    totalScore += GetScore(str, vanHoaAm.split(","));
    totalScore += GetScore(str, banNguyenAm.split(","));
    totalScore += GetScore(str, phuam.split(","));

    return totalScore;
  }

  public static int GetScore(String str, String[] checker) {
    int score = 0;
    for (int i = 0; i < checker.length; i++) {
      if (str.indexOf(checker[i]) >= 0) {
        score += checker[i].length();
      }
    }
    return score;
  }
}

File: caesar-cipher/ver1/decryptVietnamese.js
function decrypt() {
  let ciphertext = document.getElementById("inp-ciphertext").value;
  let { plaintext, ratio } = decryptCaesarCipherVietnamese(ciphertext);
  let resultText = `${plaintext} - <span style="color:red">${ratio}%<span>`;
  document.getElementById("result").innerHTML = resultText;
}

/* 
  Ý tưởng 1 : Dựa vào các tổ hợp âm trong tiếng việt
   B1: Tạo ra 26 bản decrypt từ ciphertext đầu vào
   B2: Tìm những tổ hợp phụ âm, nguyên âm, ... của tiếng việt trong từng bản decrypt
   B3: Tính điểm (dựa trên độ dài của tổ hợp tìm được) cho từng bản decrypt
   B4: Lấy bản decrypt có số điểm cao nhất.
   */

function decryptCaesarCipherVietnamese(ct) {
  let result = "";
  let bestScore = 0;

  for (let shift = 0; shift < 26; shift++) {
    let pt = "";

    for (let i = 0; i < ct.length; i++) {
      if (ct[i] == " " || ct[i] == "." || ct[i] == ",") pt += ct[i];
      else if (ct[i] == ct[i].toUpperCase()) {
        let ch = String.fromCharCode(
          ((ct.charCodeAt(i) + (26 - shift) - 65) % 26) + 65
        );
        pt += ch;
      } else {
        let ch = String.fromCharCode(
          ((ct.charCodeAt(i) + (26 - shift) - 97) % 26) + 97
        );
        pt += ch;
      }
    }

    let score = getTotalScore(pt);
    // logs += `key=${shift} => score=${score} : ${pt}\n`;
    if (score >= bestScore) {
      bestScore = score;
      result = pt;
    }
  }

  return {
    plaintext: result,
    ratio: Math.round((bestScore / ct.length) * 100 * 100) / 100,
  };
}

// https://mltav.asn.au/vietnamese/images/documents/Van/wa-van.pdf
const phuam = "ch,gh,kh,ng,ngh,nh,ph,th,tr,gi,qu";
const banNguyenAm = "oa,oe,uy,ue";
const nguyenAm = "a,e,i,o,u,y";
const vanDonGian =
  "ic,ich,im,in,inh,ip,it,uc,un,ung,ut,um,up,ech,em,en,enh,ep,et,om,on,op,ot,ac,am,an,ang,ap,at,oc,ong,ec,eng,ach,anh";
const vanHoaAm = "eo,ao,ai,oi,eu,ia,iu,ui,uu,ua";
const vanHopAmLoai1 = "ay,au";
const vanHopAmLoai2 =
  "oac,oam,oan,oat,oang,uan,uat,uang,ieu,iec,iem,ien,iep,iet,ieng,yeu,yem,yen,uoi,uoc,uom,uon,uop,uot,uong";

function getTotalScore(s) {
  let totalScore = 0;
  let str = s.toLowerCase();
  totalScore += getScore(str, vanHopAmLoai2.split(","));
  totalScore += getScore(str, vanDonGian.split(","));
  totalScore += getScore(str, vanHopAmLoai1.split(","));
  totalScore += getScore(str, vanHoaAm.split(","));
  totalScore += getScore(str, banNguyenAm.split(","));
  totalScore += getScore(str, phuam.split(","));

  return totalScore;
}

function getScore(str, checker) {
  let score = 0;

  for (let s of checker) {
    if (str.indexOf(s) >= 0) {
      score += s.length;
    }
  }
  return score;
}

/* 
  Ý tưởng 2: Dựa vào tần suất xuất hiện
   B1: 
   */

// https://vi.wiktionary.org/wiki/Wiktionary:B%E1%BA%A3ng_t%E1%BA%A7n_s%E1%BB%91/Ti%E1%BA%BFng_Vi%E1%BB%87t
const tanSuat = "n,a,o,h,i,u,g,t,c,e,d,m,y,l,r,b,v,s,k,p,x,q";



// ============================ demo text ================================
decrypt();